Editorial updates (#241)
decentralized-identity/bbs-signature: draft-irtf-cfrg-bbs-signatures.md

: The domain separation tag.

I2OSP

> : Transform an non-negative integer to an octet string, defined in Section 4 of [@!RFC8017]. Note, the output of this operation is in big-endian order.

Reviewer: ```suggestion
: An operation that transforms a non-negative integer into an octet string, defined in Section 4 of [@!RFC8017]. Note, the output of this operation is in big-endian order.
```

OS2IP

> : Transform an octet string to an non-negative integer, defined in Section 4 of [@!RFC8017]. Note, the input of this operation must be in big-endian order.

Reviewer: ```suggestion
: An operation that transforms a octet string into an non-negative integer, defined in Section 4 of [@!RFC8017]. Note, the input of this operation must be in big-endian order.
```

In definition of this signature scheme there are two possible variations based upon the sub-group selection, namely where public keys are defined in G2 and signatures in G1 OR the opposite where public keys are defined in G1 and signatures in G2. Some pairing cryptography based digital signature schemes such as [@I-D.irtf-cfrg-bls-signature] elect to allow for both variations, because they optimize for different things. However, in the case of this scheme, due to the operations involved in both signature and proof generation being computational in-efficient when performed in G2 and in the pursuit of simplicity, the scheme is limited to a construction where public keys are in G2 and signatures in G1.

### Messages

> Each of the core operations of the BBS signature scheme, expect the inputted messages to be scalar values within a given range (specifically 1 and r-1, where r is the prime order of the G1 and G2 subgroups, defined by each ciphersuite, see [Notation](#notation)). There are multiple ways to transform a message from an octet string to a scalar value. This document defines the `MapMessageToScalarAsHash` operation, which hashes an octet string to a scalar (see [MapMessageToScalarAsHash](#mapmessagetoscalarashash)). An application can use a different `MapMessageToScalar` operation, but it MUST be clearly and unambiguously defined, for all parties involved. Before using the core operations, all messages MUST be mapped to their respective scalars using the same operation. The defined [MapMessageToScalarAsHash](#mapmessagetoscalarashash) is the RECOMMENDED way of mapping octet strings to scalar values.

Reviewer 2: ```suggestion
Each of the core operations of the BBS signature scheme expect the inputted messages to be scalar values within a given range (specifically 1 and r-1, where r is the prime order of the G1 and G2 subgroups, defined by each ciphersuite, see [Notation](#notation)). There are multiple ways to transform a message from an octet string to a scalar value. This document defines the `MapMessageToScalarAsHash` operation, which hashes an octet string to a scalar (see [MapMessageToScalarAsHash](#mapmessagetoscalarashash)). An application can use a different `MapMessageToScalar` operation, but it MUST be clearly and unambiguously defined, for all parties involved. Before using the core operations, all messages MUST be mapped to their respective scalars using the same operation. The defined [MapMessageToScalarAsHash](#mapmessagetoscalarashash) is the RECOMMENDED way of mapping octet strings to scalar values.
```

- Points in `G*` will be serialized using the `point_to_octets_g*` implementation for a particular ciphersuite.
- Non-negative integers will be serialized using `I2OSP` with an output length of 8 bytes.
- Scalars will be serialized using `I2OSP` with a constant output length defined by a particular ciphersuite.

> Variable-length octet strings will be prepended with an integer value representing the number of bytes in the string. This length is encoded to octets using `I2OSP` with an output length of 8 bytes. For example, the octet string `0x14d` is encoded as `0x0000000000000002014d`. If the length of the octet string exceeds `2^64 - 1`, the octet string must be rejected.

Author: ```suggestion
```

Definitions:

- seed_dst, octet string representing the domain separation tag:

>             ciphersuite_id || "SIG_GENERATOR_SEED_," where

Author: ```suggestion
            ciphersuite_id || "SIG_GENERATOR_SEED_" where
```

            ciphersuite_id is defined by the ciphersuite and

>             "SIG_GENERATOR_SEED_" is an ASCII string comprising of 19

Author: ```suggestion
            "SIG_GENERATOR_SEED_" is an ASCII string comprised of 19
```

            bytes.
- generator_dst, octet string representing the domain separation tag:
                 ciphersuite_id || "SIG_GENERATOR_DST_", where
                 ciphersuite_id is defined by the ciphersuite and

>                  "SIG_GENERATOR_DST_" is an ASCII string comprising of

Author: ```suggestion
                 "SIG_GENERATOR_DST_" is an ASCII string comprised of
```

### MapMessageToScalarAsHash

> This operation takes an input message and maps it to a scalar value via a cryptographic hash function for the given curve. The operation takes also as an optional input a domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuite\_id || "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_", where ciphersuite\_id is the ASCII string representing the unique ID of the ciphersuite "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_" is an ASCII string comprising of 26 bytes.

Author: ```suggestion
This operation takes an input message and maps it to a scalar value via a cryptographic hash function for the given curve. The operation takes also as an optional input a domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuite\_id || "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_", where ciphersuite\_id is the ASCII string representing the unique ID of the ciphersuite "MAP\_MSG\_TO\_SCALAR\_AS\_HASH\_" is an ASCII string comprised of 26 bytes.
```

This operation makes use of expand\_message defined in [@!I-D.irtf-cfrg-hash-to-curve], in a similar way used by the hash\_to\_field operation of Section 5 from the same document (with the additional checks for getting a scalar that is 0). If an implementer wants to use hash\_to\_field instead, they MUST use the multiplicative group of integers mod r (Fr), as the target group (F). Note however, that the hash\_to\_curve document, makes use of hash\_to\_field with the target group being the multiplicative group of integers mod p (Fp). For this reason, we don’t directly use hash\_to\_field here, rather we define a similar operation (hash\_to\_scalar), making direct use of the expand\_message function, that will be defined by the hash-to-curve suite used (i.e., either expand\_message\_xmd or expand\_message\_xof). If someone also has a hash\_to\_field implementation available, with the target group been Fr, they can use this instead (adding the check for a scalar been 0).

> The operation takes as input an octet string representing the message to hash (msg), the number of the scalars to return (count) as well as an optional domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuit\_id || "H2S\_", where ciphersuite\_id is the octet string representing the unique ID of the ciphersuite and "H2S_" is an ASCII string comprising of 4 bytes.

Author: ```suggestion
The operation takes as input an octet string representing the message to hash (msg), the number of the scalars to return (count) as well as an optional domain separation tag (dst). If a dst is not supplied, its value MUST default to the octet string returned from ciphersuit\_id || "H2S\_", where ciphersuite\_id is the octet string representing the unique ID of the ciphersuite and "H2S_" is an ASCII string comprised of 4 bytes.
```

**Generator parameters**:

> - generator\_seed: A global seed value of "BBS\_BLS12381G1\_XOF:SHAKE-256\_SSWU\_RO\_MESSAGE\_GENERATOR\_SEED" (i.e., an ASCII string comprising of 59 bytes) which is used by the [create_generators](#generator-point-computation) operation to compute the required set of message generators.

Reviewer 2: ```suggestion
- generator\_seed: A global seed value of "BBS\_BLS12381G1\_XOF:SHAKE-256\_SSWU\_RO\_MESSAGE\_GENERATOR\_SEED" (i.e., an ASCII string comprised of 59 bytes) which is used by the [create_generators](#generator-point-computation) operation to compute the required set of message generators.
```

**Generator parameters**:

> - generator\_seed: A global seed value of "BBS\_BLS12381G1\_XMD:SHA-256\_SSWU\_RO\_MESSAGE\_GENERATOR\_SEED" (i.e., an ASCII string comprising of 57 bytes) which is used by the [create_generators](#generator-point-computation) operation to compute the required set of message generators.

Reviewer 2: ```suggestion
- generator\_seed: A global seed value of "BBS\_BLS12381G1\_XMD:SHA-256\_SSWU\_RO\_MESSAGE\_GENERATOR\_SEED" (i.e., an ASCII string comprised of 57 bytes) which is used by the [create_generators](#generator-point-computation) operation to compute the required set of message generators.
```

# Test Vectors

The following section details a basic set of test vectors that can be used to confirm an implementations correctness

> **NOTE** All binary data below are represented as octet strings in big endian order, encoded in hexadecimal format.

Reviewer: ```suggestion
**NOTE** All binary data below is represented as octet strings in big endian order, encoded in hexadecimal format.
```
